### PR TITLE
Make memcached a statefulset, add anti-affinity and podDisruptionBudget

### PR DESCRIPTION
**Pull request by dhilipkumars**
1) Make it statefulset
2) Use alpine images
3) Add pod disruption budge
4) Insert soft Anti-affinity for better distribution.

### CODE REVIEW

File: stable/memcached/templates/pdb.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: policy/v1beta1
```
Comment (lachie83) [inline stable/memcached/templates/pdb.yaml]: Please remove line end characters

File: stable/memcached/Chart.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 name: memcached
-version: 0.4.1
+version: 0.5.1
```
Comment (prydonius) [inline stable/memcached/Chart.yaml]: 0.5.0

File: stable/memcached/templates/statefulset.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
-apiVersion: extensions/v1beta1
-kind: Deployment
+apiVersion: apps/v1beta1
+kind: StatefulSet
```
Comment (lachie83) [inline stable/memcached/templates/statefulset.yaml:2]: How about we add the option to have some persistent storage given that you are moving this to a statefulset?

File: stable/memcached/templates/pdb.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ template "fullname" . }}
```
Comment (prydonius) [inline stable/memcached/templates/pdb.yaml]: This resource should have the app, chart, release and heritage labels as in the other resources.

File: stable/memcached/templates/svc.yaml [hunk 5]
```diff
@@ -8,6 +8,7 @@ metadata:
     release: "{{ .Release.Name }}"
     heritage: "{{ .Release.Service }}"
 spec:
+  clusterIP: None
```
Comment (lachie83) [inline stable/memcached/templates/svc.yaml:11]: What's the purpose of this? Is this required for Statefulset?

**Review by lachie83 — CHANGES_REQUESTED** [review]

**Review by prydonius — CHANGES_REQUESTED** [review]
Thanks @dhilipkumars

File: stable/memcached/templates/statefulset.yaml (hunk 3 above)
Comment (dhilipkumars) [reply stable/memcached/templates/statefulset.yaml:2]: AFAIK.  I dont think we use memcachd for disk persistence, popular usecase is 
```
value = Cache.Lookup(key)
if key did not exist then
  value =  DB.Read(key)
  cache.Set(key, value)
fi
```
we generally re-populate the cache, so if we persist and repopulate on re-starts there is no gaurentee that the cache is the latest.

File: stable/memcached/templates/svc.yaml (hunk 5 above)
Comment (dhilipkumars) [reply stable/memcached/templates/svc.yaml:11]: Yes statefulset requires Headlessservices and AKIK [this ](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services)is the way to create headless services.

File: stable/memcached/Chart.yaml (hunk 2 above)
Comment (dhilipkumars) [reply stable/memcached/Chart.yaml]: done

File: stable/memcached/templates/pdb.yaml (hunk 1 above)
Comment (dhilipkumars) [reply stable/memcached/templates/pdb.yaml]: done, and sorry

File: stable/memcached/templates/pdb.yaml (hunk 4 above)
Comment (dhilipkumars) [reply stable/memcached/templates/pdb.yaml]: added them.

**Comment by dhilipkumars** [comment]
@prydonius and @lachie83 Thanks for reviewing PTAL now.

**Comment by seanknox** [comment]
@k8s-bot ok to test

**Comment by dhilipkumars** [comment]
@seanknox 
```
W0503 20:27:25.074] Error: release memcach-1468 failed: error validating "": error validating data: found invalid field affinity for v1.PodSpec
```
Are these charts tested against older kubernetes like 1.5?  Affinity moved to podspec from 1.6.  I think that is the error.

**Comment by seanknox** [comment]
I believe charts test against v1.5 only at the moment.

**Comment by dhilipkumars** [comment]
How do we handle this situation? Shouldn't it be upgraded to 1.6?

**Comment by dhilipkumars** [comment]
Hi @seanknox, i was told that the CI is now upgraded to 1.6, how do i trigger a test again?

**Comment by seanknox** [comment]
@k8s-bot ok to test

**Comment by seanknox** [comment]
Thanks @dhilipkumars

**Comment by prydonius** [comment]
Switching from Deployment to StatefulSet is a breaking change so I would bump this chart to 1.0.0. Other than that, this lgtm - thanks @dhilipkumars!

**Comment by dhilipkumars** [comment]
@prydonius Thanks for the suggestion it is done now.  PTAL

**Review by prydonius — APPROVED** [review]

**Comment by prydonius** [comment]
Thanks @dhilipkumars!

**Pull request merged**